### ISSUE TITLE
Auto-scaling issue when interactively removing EELS background

### ISSUE THREAD

**Issue by jat255** [hyperspy/hyperspy#562]
When trying to remove the background with the GUI, the scale of the plot will sometimes change drastically such that I will not be able to see the data at all. I can either close and restart the process, or drag the data into view with the window controls. I've figured out that this behavior is dependent on where you click in the spectrum. If you click right on the line, it will work fine. If you click far from the line (vertically), then the scale will get all out of sorts. See this video for example, where the first time I click near the line, and then the next time I click far from it:

![remove_background_scale](https://cloud.githubusercontent.com/assets/1278301/7966792/5c49d4fa-09f4-11e5-9df3-1d02b89c2126.gif)

**Comment by jat255** [comment]
I'll also note that this doesn't happen if I change the background type to 'offset' first (even if I click and drag far from the spectrum), so I'm assuming this has something to do with the fitting hitting extreme data values, which then throws the scale out of sorts:

![remove_background_offset](https://cloud.githubusercontent.com/assets/1278301/7967193/9428d720-09f6-11e5-8f9f-5dfbd63e7ba1.gif)

**Comment by vidartf** [comment]
I often have this problem too. #411 reduced the occurrence of this, but it is still present. I suspect that what happens is this:
When a region is first dragged out, maybe only two data points are included. If the data is noisy, those two points might then correspond to a very extreme powerlaw. As that powerlaw is plotted, the ylimits are expanded to accommodate this extreme plot, which completely screws up the plot. A possible solution would be to continually fit the ylimits as we plot (i.e. shrink them as applicable, as expansions is already handled by MPL). Maybe there is an MPL setting/function?

**Comment by jat255** [comment]
@vidartf I thought that that was what is happening as well. Although, if that's the case, I am not sure why dragging far from the plotted lines causes the behavior, while dragging close to the curve does not...

**Comment by vidartf** [comment]
Are you sure the x-coordinate is the same every time? It could be a random process due to the noise (not every pair of points will give an extreme power law).

**Comment by jat255** [comment]
Hmm, it appears you are right. 

[This](https://dl.dropboxusercontent.com/u/10983031/example.hdf5) is the file that I'm using to test: It appears that regardless of where on the y-axis I click, if I click at x = 72.3521 and drag right, it will always cause the extreme scaling behavior.

Likewise, at x=79.0115 (and dragging right), regardless of where on the y-axis I click, it does not put the scale out of sorts. I guess maybe it was just coincidence that it was happening depending on where I clicked on the y-axis.

**Comment by jat255** [comment]
Also, changing to `offset` for the fitting type means I don't see the crazy scale at either point, so I think you are right that it has to do with the fit. I could envision a couple solutions:

1. Explicitly define the y-axis limits based on the min and max of the plot (like you were suggesting, I think)
2. Prevent the algorithm from fitting until the mouse is released (although this would lose the interactivity of the process that is really nice to have)
3. Prevent the algorithm from fitting if the two end points are within a certain number of pixels. Perhaps explicitly saying that the left and right edges of the window must be at least 5 channels apart or something would be enough. 
4. Operate on a smoothed version of the data to prevent the noise from having an impact. The real data could still be presented to the user, but behind the scenes a smoothed version could be used for the fitting.

**Comment by vidartf** [comment]
> Explicitly define the y-axis limits based on the min and max of the plot (like you were suggesting, I think)

Basically yes: Either lock the limits while fitting, or make sure that the ylimits can shrink again if they have been expanded (the extreme power laws normally only pop up early in the area definition).

**Comment by francisco-dlp** [comment]
For me there is something mysterious going on here. In principle the background line (i.e. the blue line that represents the fit to the background) shouldn't autoscale the y-axis because its `autoscale` attribute is set to `False` [this is the code](https://github.com/hyperspy/hyperspy/blob/master/hyperspy/gui/egerton_quantification.py#L108) . Therefore, it shouldn't be necessary to force the y-axis to freeze the limits. And indeed, if changing the background window didn't trigger autoscale at the very start of the dragging process, later it doesn't trigger it anymore i.e. it's possible to select a background window that takes the line out of the y-axis limits and the y-axis doesn't autoscale. Indeed, if one if fast enough at expanding the slider even at x = 72.3521 the y-axis doesn't autoscale. Does anybody has any clue of what is going on here?

**Comment by vidartf** [comment]
> In principle the background line (i.e. the blue line that represents the fit to the background) shouldn't autoscale the y-axis because its autoscale attribute is set to False this is the code . 

Just quickly played around with this: The background fit plot does not update the limits when updating, however, it seems that the limits are updated for the very first plot of the line! Normally, this simply means to include the y-axis 0 value, but if the power law goes the wrong way it will include some very high values. This is also why the limits are not reset later. To fix this simply switch line [107 and 108](https://github.com/hyperspy/hyperspy/blob/master/hyperspy/gui/egerton_quantification.py#L107-108): 

```python
# This
        self.signal._plot.signal_plot.add_line(self.bg_line)
        self.bg_line.autoscale = False

# Becomes:
        self.bg_line.autoscale = False
        self.signal._plot.signal_plot.add_line(self.bg_line)
```

**Comment by francisco-dlp** [comment]
Fixed by #577.
